addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-download.go

```diff
@@ -86,24 +133,135 @@ func getBucketObjects(sess *session.Session, config *mountConfiguration, concurr
+
 	for truncatedListing {
 		resp, err := svc.ListObjectsV2(query)
 
 		if err != nil {
 			log.Println("Failed to list objects: ", err)
+			// 30 seconds backoff
 			time.Sleep(time.Duration(30) * time.Second)
 			continue
 		}
-		getObjectsAll(resp, sess, config, concurrency, stats, debug)
+
+		listObjectResponses = append(listObjectResponses, resp)
+		downloadAllObjects(resp, sess, config, concurrency, stats, debug)
 
 		query.ContinuationToken = resp.NextContinuationToken
 		truncatedListing = *resp.IsTruncated
 	}
 
+	err := deleteLocalFilesNotInS3(listObjectResponses, config, debug)
+	if err != nil {
+		fmt.Println("Error: ", err)
+	}
+
+	stats.end = time.Now()
 	return stats
 }
 
-func getObjectsAll(
+func deleteLocalFilesNotInS3(listObjectResponses []*s3.ListObjectsV2Output, config *mountConfiguration, debug bool) error {
+	destination := config.destination
+	prefix := config.prefix
+
+	findInS3 := func(path string) *s3.Object {
+		for _, listObjectResponse := range listObjectResponses {
+			for _, item := range listObjectResponse.Contents {
+				// Strip the base s3 prefix
+				destFilename := strings.TrimPrefix(*item.Key, prefix)
+				destFilePath := filepath.Join(destination, destFilename)
+				if path == destFilePath {
+					return item
+				}
+			}
+		}
+		return nil
+	}
+
+	walkerFn := func(path string, info os.FileInfo, err error) error {
+		// Don't do anything if there was any error during walking the file tree
+		if err != nil {
+			fmt.Printf("\nError walking the file tree: \"%s\". Error: %v\n", path, err)
+			return nil
+		}
+		if info.Mode().IsDir() {
+			// Ignore directories
+			return nil
+		}
+
+		fileInS3 := findInS3(path)
+		if fileInS3 == nil {
+			// file NOT in S3 but is in local file system -- DELETE from local file system
+			if debug {
+				fmt.Printf("\n\nFile '%s' removed from S3 so deleting it from local file system\n\n", path)
+			}
+			error := os.Remove(path)
+			if error == nil {
+				recordFileDeletionFromLocal(path, config)
+			} else {
+				fmt.Printf("\nError deleting file: \"%s\". Error: %v\n", path, error)
+			}
+		}
+		return nil
+	}
+
+	err := filepath.Walk(destination, walkerFn)
+	return err
+}
+
+func setupRecurringDownloads(wg *sync.WaitGroup, sess *session.Session, config *mountConfiguration, concurrency int, debug bool, downloadInterval int, stopRecurringDownloadsAfter int) {
+	// Increment wait group counter everytime we spawn recurring downloads thread to make sure
+	// the caller (main) can wait
+	wg.Add(1)
+
+	statsCh := make(chan *downloadStats, 50)
+
+	continueRecurringDownloads := true
+	setupStartTime := time.Now()
+
+	// Kick off thread for recurring download for this mount configuration
+	// This thread will push download stats to stats channel and the reporter thread will receive stats from the
+	// stats channel and report (print) the stats
+	go func() {
+		for continueRecurringDownloads {
+			stats := syncS3ToLocal(sess, config, concurrency, debug)
```

Comment:
I know you handle probably all the errors in sync method. I was wondering though if we should have some error handling here. Just for my education- in case sync method throws an error, is this thread still active or just dies out silently?

Reply:
In Go, there is no concept of "throwing an error". Since Go functions can return multiple values, the errors are just returned along with other values just like regular return values. Unless we explicitly read error and exit, returning an error does not automatically exit the thread.

Reply:
Thanks, good to know. Thanks for clarifying